### PR TITLE
Enabled strings::Join(*).

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

File: Bricks/cerealize/cerealize.h [hunk 1]
```diff
@@ -44,72 +44,29 @@ SOFTWARE.
...
-  constexpr static bool value = true;
-};
-// Microsoft Visual Studio compiler is strict with overloads,
-// explicitly exclude string-related types from cereal-based implementations.
-template <typename TOP_LEVEL_T>
-struct is_string_type {
-  constexpr static bool value = is_string_type_impl<decay<TOP_LEVEL_T>>::value;
-};
+// Explicitly exclude string-related types from cereal-based implementations.
+// (Had troubles with Microsoft Visual Studio compiler matching too much, likely, correctly. -- D.K.)
```
Comment (mzhurovich) [inline Bricks/cerealize/cerealize.h]: This comment looks dangling. Refer to `strings/is_string_type.h`?

Comment (dkorolev) [inline Bricks/cerealize/cerealize.h]: Hmm, I specifically kept it here since it has to do with Cereal, not with strings. Maybe remove it altogether?

Comment (mzhurovich) [inline Bricks/cerealize/cerealize.h]: Let's remove it then. It really looks unobvious :)

Comment (dkorolev) [inline Bricks/cerealize/cerealize.h]: Done, another couple mins for SFINAE.

**Comment by mzhurovich** [comment]
BTW, how about default empty separator for `Join`?

**Comment by dkorolev** [comment]
I'd be against a default empty separator. We're not Perl :-)

**Comment by dkorolev** [comment]
@mzhurovich : Pushed the change in response to your comment and added and tested SFINAE magic. PTAL!

File: Bricks/strings/join.h [hunk 2]
```diff
@@ -59,26 +61,84 @@ inline size_t StringLengthOrOneForChar(T&& param) {
...
+  enum { value = IsContainerOfStringsImpl<T, HasBegin<T>::CompileTimeCheck(0)>::value };
+};
+
+}  // namespace sfinae
+
+template <typename T_CONTAINER, typename T_SEPARATOR>
+void OptionallyReserveOutputBuffer(std::string& output,
+                                   const T_CONTAINER& components,
+                                   T_SEPARATOR&& separator) {
+  // Note: this implementation does not to `reserve()` for chars, the length of which is always known to be 1.
```
Comment (mzhurovich) [inline Bricks/strings/join.h]: s/not to `reserve()`/not `reserve()`/

Comment (dkorolev) [inline Bricks/strings/join.h]: It was `do`. Fixing.

Comment (dkorolev) [inline Bricks/strings/join.h]: Done in b4e0259.

File: Bricks/strings/join.h [hunk 3]
```diff
@@ -59,26 +61,84 @@ inline size_t StringLengthOrOneForChar(T&& param) {
...
+    length += impl::StringLengthOrOneForChar(separator) * (components.size() - 1);
+    output.reserve(length);
+  }
+};
+
+namespace sfinae {
+
+template <typename T>
+struct HasBegin {
+  template <typename... XS>
```
Comment (mzhurovich) [inline Bricks/strings/join.h]: Why do you need this here?

Comment (dkorolev) [inline Bricks/strings/join.h]: :+1: We don't need it here. Removed in 0908ab4. Good eye once again!

File: Bricks/strings/util.h [hunk 4]
```diff
@@ -25,33 +25,56 @@ SOFTWARE.
...
 
 namespace bricks {
 namespace strings {
 
+template <typename DECAYED_T>
+struct ToStringImpl {
+  template <typename T>
+  static std::string ToString(T&& something) {
+    static_assert(std::is_same<bricks::decay<T>, DECAYED_T>::value, "");
+    return std::to_string(std::forward<T>(something));
```
Comment (mzhurovich) [inline Bricks/strings/util.h]: `std::forward` is overkill here - `std::to_string` accepts only a number of numeric types.

Comment (dkorolev) [inline Bricks/strings/util.h]: Looking into, ETA 3 min.

File: Bricks/strings/util.h [hunk 5]
```diff
@@ -25,33 +25,56 @@ SOFTWARE.
...
+};
+
+template <>
+struct ToStringImpl<char*> {  // Decayed type in template parameters list.
+  static std::string ToString(const char* string) { return string; }
+};
+
+template <int N>
+struct ToStringImpl<char[N]> {  // Decayed type in template parameters list.
+  static std::string ToString(const char chunk[N]) {
```
Comment (mzhurovich) [inline Bricks/strings/util.h]: Maybe rename `chunk` to `string`? IMHO it's not clear from the name that `chunk`must be null-terminated.

Comment (dkorolev) [inline Bricks/strings/util.h]: Done in 8af4005

File: Bricks/strings/util.h (hunk 4 above)
Comment (dkorolev) [inline Bricks/strings/util.h]: Great catch, done using `is_pod<>` in 8af4005.

**Comment by dkorolev** [comment]
@mzhurovich : PTAL.

**Comment by mzhurovich** [comment]
LGTM!
Merging.

**Comment by dkorolev** [comment]
Yay! Thanks Max!

Putting https://github.com/c5t/Current/pull/203 into shape now, with this and yours changes.

**Pull request merged**